File: sdk/java-sdk/src/test/java/com/evernym/verity/sdk/protocols/UpdateConfigsTest.java
package com.evernym.verity.sdk.protocols;

import com.evernym.verity.sdk.TestHelpers;
import com.evernym.verity.sdk.exceptions.VerityException;
import com.evernym.verity.sdk.protocols.updateconfigs.UpdateConfigs;
import com.evernym.verity.sdk.protocols.updateconfigs.v0_6.UpdateConfigsV0_6;
import com.evernym.verity.sdk.utils.Context;
import com.evernym.verity.sdk.utils.Util;
import org.json.JSONArray;
import org.json.JSONObject;
import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

import static com.evernym.verity.sdk.TestHelpers.unpackForwardMessage;
import static org.junit.Assert.*;

public class UpdateConfigsTest {

    private final String name = "Name1";
    private final String logoUrl = "http://logo.url";

    @Test
    public void testGetMessageType() {
        UpdateConfigsV0_6 testProtocol = UpdateConfigs.v0_6(name, logoUrl);
        String msgName = "msg name";
        assertEquals(Util.getMessageType(
                Util.EVERNYM_MSG_QUALIFIER,
                testProtocol.family(),
                testProtocol.version(),
                msgName
        ), testProtocol.messageType(msgName));
    }

    @Test
    public void testConstructorWithAttr() throws VerityException {
        Context context = TestHelpers.getContext();
        UpdateConfigsV0_6 testProtocol = UpdateConfigs.v0_6(name, logoUrl);

        JSONObject msg = testProtocol.updateMsg(context);
        testUpdateMsgMessages(msg);

        JSONObject msg2 = testProtocol.statusMsg(context);
        testStatusMsg(msg2);
    }

    @Test
    public void testConstructorWithoutAttr() throws VerityException {
        Context context = TestHelpers.getContext();
        UpdateConfigsV0_6 testProtocol = UpdateConfigs.v0_6();

        JSONObject msg2 = testProtocol.statusMsg(context);
        testStatusMsg(msg2);
    }

    private void testUpdateMsgMessages(JSONObject requestMsg) {
        Map<String, String> expectedConfigs = new HashMap<>();
        expectedConfigs.put("name", this.name);
        expectedConfigs.put("logoUrl", this.logoUrl);

        assertEquals(
                "did:sov:123456789abcdefghi1234;spec/update-configs/0.6/update",
                requestMsg.getString("@type")
        );
        assertNotNull(requestMsg.getString("@id"));
        assertEquals(
                parseConfigs(requestMsg.getJSONArray("configs")),
                expectedConfigs
        );
    }

    private void testStatusMsg(JSONObject statusMsg) {
        assertEquals(
                "did:sov:123456789abcdefghi1234;spec/update-configs/0.6/get-status",
                statusMsg.getString("@type"));
        assertNotNull(statusMsg.getString("@id"));
    }

    @Test
    public void testUpdate() throws Exception {
        Context context = null;
        try {
            context = TestHelpers.getContext();
            UpdateConfigsV0_6 testProtocol = UpdateConfigs.v0_6(name, logoUrl);
            byte [] message = testProtocol.updateMsgPacked(context);
            JSONObject unpackedMessage = unpackForwardMessage(context, message);
            testUpdateMsgMessages(unpackedMessage);
        } catch(Exception e) {
            e.printStackTrace();
            fail();
        } finally {
            TestHelpers.cleanup(context);
        }
    }

    @Test
    public void testGetStatus() throws Exception {
        Context context = null;
        try {
            context = TestHelpers.getContext();
            UpdateConfigsV0_6 testProtocol = UpdateConfigs.v0_6();
            byte [] message = testProtocol.statusMsgPacked(context);
            JSONObject unpackedMessage = unpackForwardMessage(context, message);
            testStatusMsg(unpackedMessage);
        } catch(Exception e) {
            e.printStackTrace();
            fail();
        } finally {
            TestHelpers.cleanup(context);
        }
    }

    private Map parseConfigs(JSONArray configsJson) {
        Map<String, String> configs = new HashMap<>();
        for(int i = 0; i < configsJson.length(); i++) {
            JSONObject c = configsJson.getJSONObject(i);
            configs.put(c.getString("name"), c.getString("value"));
        }

        return configs;
    }
}

File: sdk/dotnet-sdk/src/VeritySDK.Test/Protocols/UpdateConfigs/UpdateConfigsTest.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Collections.Generic;
using System.Json;
using System.Linq;
using VeritySDK.Protocols.UpdateConfigs;
using VeritySDK.Utils;

namespace VeritySDK.Test
{
    [TestClass]
    public class UpdateConfigsTest : TestBase
    {
        private string name = "Name1";
        private string logoUrl = "http://logo.url";

        [TestMethod]
        public void testGetMessageType()
        {
            UpdateConfigsV0_6 testProtocol = UpdateConfigs.v0_6(name, logoUrl);
            string msgName = "msg name";
            Assert.AreEqual(Util.getMessageType(
                    Util.EVERNYM_MSG_QUALIFIER,
                    testProtocol.family(),
                    testProtocol.version(),
                    msgName
            ), testProtocol.messageType(msgName));
        }

        [TestMethod]
        public void testGetThreadId()
        {
            UpdateConfigsV0_6 testProtocol = UpdateConfigs.v0_6(name, logoUrl);
            Assert.IsNotNull(testProtocol.getThreadId());
        }

        [TestMethod]
        public void testConstructorWithAttr()
        {
            Context context = TestHelpers.getContext();
            UpdateConfigsV0_6 testProtocol = UpdateConfigs.v0_6(name, logoUrl);

            JsonObject msg = testProtocol.updateMsg(context);
            testUpdateMsgMessages(msg);

            JsonObject msg2 = testProtocol.statusMsg(context);
            testStatusMsg(msg2);
        }

        [TestMethod]
        public void testConstructorWithoutAttr()
        {
            Context context = TestHelpers.getContext();
            UpdateConfigsV0_6 testProtocol = UpdateConfigs.v0_6();

            JsonObject msg2 = testProtocol.statusMsg(context);
            testStatusMsg(msg2);
        }

        private void testUpdateMsgMessages(JsonObject requestMsg)
        {
            Dictionary<string, string> expectedConfigs = new Dictionary<string, string>();
            expectedConfigs.Add("name", this.name);
            expectedConfigs.Add("logoUrl", this.logoUrl);

            Assert.AreEqual(
                    "did:sov:123456789abcdefghi1234;spec/update-configs/0.6/update",
                    requestMsg.getAsString("@type")
            );
            Assert.IsNotNull(requestMsg.getAsString("@id"));
            Assert.IsTrue(parseConfigs(requestMsg.getAsJsonArray("configs")).EquivalentTo(expectedConfigs));
        }

        private void testStatusMsg(JsonObject statusMsg)
        {
            Assert.AreEqual(
                    "did:sov:123456789abcdefghi1234;spec/update-configs/0.6/get-status",
                    statusMsg.getAsString("@type"));
            Assert.IsNotNull(statusMsg.getAsString("@id"));
        }

        [TestMethod]
        public void testUpdate()
        {
            withContext(context =>
            {
                UpdateConfigsV0_6 testProtocol = UpdateConfigs.v0_6(name, logoUrl);
                byte[] message = testProtocol.updateMsgPacked(context);
                JsonObject unpackedMessage = TestHelpers.unpackForwardMessage(context, message);
                testUpdateMsgMessages(unpackedMessage);
            });
        }

        [TestMethod]
        public void testGetStatus()
        {
            withContext(context =>
            {
                UpdateConfigsV0_6 testProtocol = UpdateConfigs.v0_6();
                byte[] message = testProtocol.statusMsgPacked(context);
                JsonObject unpackedMessage = TestHelpers.unpackForwardMessage(context, message);
                testStatusMsg(unpackedMessage);
            });
        }

        private Dictionary<string, string> parseConfigs(JsonArray configsJson)
        {
            Dictionary<string, string> configs = new Dictionary<string, string>();
            for (int i = 0; i < configsJson.Count; i++)
            {
                JsonObject c = configsJson.ElementAt(i) as JsonObject;
                configs.Add(c.getAsString("name"), c.getAsString("value"));
            }

            return configs;
        }
    }
}
